barretenberg/cpp/src/barretenberg/crypto/merkle_tree/lmdb_store/lmdb_tree_store.hpp

```diff
@@ -63,39 +65,90 @@ struct BlockIndexPayload {
     }
 
     void delete_block(const block_number_t& blockNumber)
     {
+        // Shouldn't be possible, but no need to do anything here
         if (blockNumbers.empty()) {
             return;
         }
-        // shuffle the block number down, removing the one we want to remove and then pop the end item
-        auto it = std::lower_bound(blockNumbers.begin(), blockNumbers.end(), blockNumber);
-        if (it == blockNumbers.end()) {
-            // The block was not found, we can return
+
+        // If the size is 1, the blocknumber must match that in index 0, if it does remove it
+        if (blockNumbers.size() == 1) {
+            if (blockNumbers[0] == blockNumber) {
+                blockNumbers.pop_back();
+            }
             return;
         }
-        // It could be a block higher than the one we are looking for
-        if (*it != blockNumber) {
+
+        // we have 2 entries, we must verify that the block number is equal to the item in index 1
+        if (blockNumbers[1] != blockNumber) {
+            throw std::runtime_error(format("Unable to delete block number ",
+                                            blockNumber,
+                                            " for retrieval by index, current max block number at that index: ",
+                                            blockNumbers[1]));
+        }
+        // It is equal, decrement it, we know that the new block number must have been added previously
+        --blockNumbers[1];
+
+        // We have modified the high block. If it is now equal to the low block then pop it
+        if (blockNumbers[0] == blockNumbers[1]) {
+            blockNumbers.pop_back();
+        }
+    }
+
+    void add_block(const block_number_t& blockNumber)
+    {
+        // If empty, just add the block number
+        if (blockNumbers.empty()) {
+            blockNumbers.emplace_back(blockNumber);
+            return;
+        }
+
+        // If the size is 1, then we must be adding the block 1 larger than that in index 0
+        if (blockNumbers.size() == 1) {
+            if (blockNumber != blockNumbers[0] + 1) {
+                // We can't accept a block number for this index that does not immediately follow the block before
+                throw std::runtime_error(format("Unable to store block number ",
+                                                blockNumber,
+                                                " for retrieval by index, current max block number at that index: ",
+                                                blockNumbers[0]));
+            }
+            blockNumbers.emplace_back(blockNumber);
             return;
         }
-        // we have found our block, shuffle blocks after this one down
-        auto readIt = it + 1;
-        while (readIt != blockNumbers.end()) {
-            *it++ = *readIt++;
+
+        // Size must be 2 here, if larger, this is an error
+        if (blockNumbers.size() != 2) {
```

Comment:
WDYT of changing this to explicit `interval_start, interval_end` member variables?